### PR TITLE
CLOUDSTACK-8272: Python based file-lock free password server implementat...

### PR DESCRIPTION
**Pull request by bhaisaab**
Major changes:

- VRs are single CPU, so Threading based implementation favoured than Forking based
- Implements a Python based password server that does not use file based locks
- Saving password mechanism is provided by using secure token only to VR (localhost)
- Old serve_password implementation is removed
- Runs with Python 2.6+ with no external dependencies
- Locks used within threads for extra safety

The aim for this is to improve password server by relying on a threading based approach to minimize RAM usage and at the same time avoid using file based lock. For 4.5.1 and 4.6.0

### CODE REVIEW

**Comment by vincentbernat** [comment]
Some remarks:

 - Use `with lock:` instead of `lock.acquire()` then `try/finally` (compatible with Python 2.6)
 - Use `with file(...) as f:` (idem)
 - `global passMap` and `global lock` are useless since you don't do direct assignments
 - `getPassword()` could be `passMap.get(..., None)`

File: systemvm/patches/debian/config/opt/cloud/bin/passwd_server_ip.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+def getPassword(ip):
+    global passMap
+    if ip in passMap:
+        return passMap[ip]
+    return None
+
+def setPassword(ip, password):
+    global passMap, lock
+    if ip == None or ip == "" or password == None or password == "":
```
Comment (brutasse) [inline systemvm/patches/debian/config/opt/cloud/bin/passwd_server_ip.py]: You could simply test for `if not ip or not password`

File: systemvm/patches/debian/config/opt/cloud/bin/passwd_server_ip.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        form = cgi.FieldStorage(
+                    fp=self.rfile,
+                    headers=self.headers,
+                    environ={'REQUEST_METHOD':'POST',
+                             'CONTENT_TYPE':self.headers['Content-Type'],
+                    })
+        self.send_response(200)
+        self.end_headers()
+        clientAddress = self.client_address[0]
+        if not (clientAddress == "localhost" or clientAddress == "127.0.0.1"):
```
Comment (brutasse) [inline systemvm/patches/debian/config/opt/cloud/bin/passwd_server_ip.py]: `if clientAddress not in ['localhost', '127.0.0.1']`

File: systemvm/patches/debian/config/opt/cloud/bin/passwd_server_ip.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            syslog.syslog("serve_password: non-localhost IP trying to save password: %s" % clientAddress)
+            self.send_response(403)
+            return
+        if "ip" not in form.keys() or "password" not in form.keys() or "token" not in form.keys() or self.headers.get('DomU_Request') != "save_password":
```
Comment (brutasse) [inline systemvm/patches/debian/config/opt/cloud/bin/passwd_server_ip.py]: With dictionaries you can directly check for membership with `if "ip" not in form`.

**Comment by bhaisaab** [comment]
Thanks @vincentbernat @brutasse for your reviews, fixing them now.

**Comment by bhaisaab** [comment]
In a general discussion, should the stored passwords be made expirable; i.e. after a certain interval if the passwords are not served they expire, or if within certain interval (say half an hour) if any password is served should be expired?

**Comment by pdion891** [comment]
@bhaisaab, can you confirm that if the password is not grap after a certain period it expired (ex: 15mins), and also if the password is grap once, it is not available anymore and have to be reset again. Looks like it is the current approach.  Does the current script use inside VMs will remain functional?

Thanks

**Comment by bhaisaab** [comment]
@vincentbernat thanks for the snippets!

@resmo in case the passwdsrvrtoken went missing (file does not exist), the script may break, the || true is just so that it return 0 (exit 0). I guess your snippet is correct as well (check and cat file), using that approach.

@pdion891 looking at the code at least on the VR side there is no password expiry period imposed by the server. I can sort of understand this, because someone could shutdown a server, do a reset password and if the VM won't get started we should store the password in the VR. But let's discuss the security aspects on what/how we should implement a self expiring cache. I propose:
1. if we have served the password at least once, let's expire it in 15-30 minutes
2. else, expire all the passwords in next 1-24 hours whether they have been served to user VMs (for resetting purposes) or not.

Comments?

**Comment by bhaisaab** [comment]
Tested against KVM works for me, closing. Merged on 4.5 and master.

Opened two ticket to track futher work on it to implement a SSL based server and IPv6 support.

**Pull request closed without merging**